writer-v2.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (C) 2018 Minio Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sio
+
+import (
+	"io"
+)
+
+type encWriterV20 struct {
+	authEncV20
+	dst io.Writer
+
+	buffer packageV20
+	offset int
+}
+
+func encryptWriterV20(dst io.Writer, config *Config) (*encWriterV20, error) {
+	ae, err := newAuthEncV20(config)
+	if err != nil {
+		return nil, err
+	}
+	return &encWriterV20{
+		authEncV20: ae,
+		dst:        dst,
+		buffer:     make(packageV20, maxPackageSize),
+	}, nil
+}
+
+func (w *encWriterV20) Write(p []byte) (n int, err error) {
+	if w.finalized {
```

Comment:
Shouldn't this be set in `Close()` for encWriterV20 ?

Reply:
`Close()` will set it. However a client can try to write more data to the writer which is a bug in the calling code. Since the `w.finalized` indicates that the last package (which is marked as last package) was written to the underlying `io.Writer`. So writing more data after `w.finalized` would mean you try that append a package to the *last* package.
The only way this can happen is:
 1. Client calls `encWriterV2.Write(data)`
 2. Client calls `encWriterV2.Close()`
 3. Client calls `encWriterV2.Write(data)` <- this will panic

Reply:
Yes but current Close() in encWriter20 is not setting `w.finalized = true`

Reply:
It does: It calls `SealFinal` -> https://github.com/minio/sio/pull/20/files#diff-3d91b17e0e42b0766b31da5ebaada786R188